FirebaseAdmin/FirebaseAdmin.IntegrationTests/Auth/TemporaryUserBuilder.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using FirebaseAdmin.Auth;
+
+namespace FirebaseAdmin.IntegrationTests.Auth
+{
+    /// <summary>
+    /// A utility for creating temporary user accounts during tests. Keeps track of the user
+    /// accounts created, and deletes them on dispose. Other user accounts created outside this
+    /// class can be marked for deletion using the <see cref="AddUid(string)"/> method. This class
+    /// deletes user accounts in an idempotent manner. Therefore it's safe to delete any of the
+    /// user accounts created by this class before this instance is disposed.
+    /// <summary>
+    public sealed class TemporaryUserBuilder : IDisposable
+    {
+        private readonly ISet<string> userIds = new HashSet<string>();
+
+        public TemporaryUserBuilder()
+        {
+            IntegrationTestUtils.EnsureDefaultApp();
+        }
+
+        public static UserRecordArgs RandomUserRecordArgs()
+        {
+            var uid = Guid.NewGuid().ToString().Replace("-", string.Empty);
+            var rand = new Random();
+            var phoneDigits = Enumerable.Range(0, 10).Select(_ => rand.Next(10));
+
+            return new UserRecordArgs()
+            {
+                Uid = uid,
+                Email = $"test{uid.Substring(0, 12)}@example.{uid.Substring(12)}.com",
+                PhoneNumber = $"+1{string.Join(string.Empty, phoneDigits)}",
+                DisplayName = "Random User",
+                PhotoUrl = "https://example.com/photo.png",
+                Password = "password",
+            };
+        }
+
+        public async Task<UserRecord> CreateRandomUserAsync()
+        {
+            return await this.CreateUserAsync(RandomUserRecordArgs());
+        }
+
+        public async Task<UserRecord> CreateUserAsync(UserRecordArgs args)
+        {
+            // Make sure we never create more than 1000 users in a single instance.
+            // This allows us to delete all user accounts with a single call to DeleteUsers().
+            // Should not ever occur in practice.
+            if (this.userIds.Count() == 1000)
```

Comment:
Is concurrency a concern? I don't think the tests run in parallel, so I would guess no. But if that's wrong, then consider a mutex.

Somewhat related, consider >= 1000 instead of ==, just to be a little extra paranoid.

Reply:
I thought about the concurrency issue. Tests in the same class are never run in parallel, so this won't be a problem unless we reuse an instance of this class across classes. But I'm inclined to just call this class *not thread safe* and ask the caller synchronize if necessary. I've updated the class summary to that effect.

Changed the userId count check as indicated.